File: DFTP.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>


// Pre-declarations -----------------------------------------------------------
void init(char**, char**, int, char**);
void print_header(char*, char*);
void get_request(char**, size_t*);
void print_help_message();
void DFTP_request(char*);


// main -----------------------------------------------------------------------
int main(int argc, char* argv[]){
	// Initialize variables
	char *host, *port;
	size_t req_size = 256;
	char* request = (char*) malloc(req_size * sizeof(char));
	init(&host, &port, argc, argv);
	
	// TODO - Calculate IP
	

	print_header(host, port);
	
	while(1){
		get_request(&request, &req_size);

		if(strcmp(request, "quit") == 0){
			break;
		}else if(strcmp(request, "help") == 0){
			print_help_message();
		}else{
			DFTP_request(request);
		}
	}
	
	puts("Goodbye.");
	free(request);
	return 0;
}


// init -----------------------------------------------------------------------
void init(char** host, char** port, int argc, char* argv[]){
	if(argc > 1){
		*host = argv[1];
	}else{
		*host = "unyama.cs.ucdavis.edu";
	}
	
	if(argc > 2){
		*port = argv[2];
	}else{
		*port = "10000";
	}
}


// print_header ---------------------------------------------------------------
void print_header(char* host, char* port){
	printf("Welcome to the simple DFTP client.\n");
	printf("Connecting to %s(Calc IP) on port %s.\n", host, port);
}


// get_request ----------------------------------------------------------------
void get_request(char** request, size_t* size){
	printf("command > ");
	getline(request, size, stdin);
	char* newline = strchr(*request, '\n');
	*newline = '\0';
}


// print_help_message ---------------------------------------------------------
void print_help_message(){
	puts("help - Prints this message");
	puts("quit - Exits the program");
	puts("GET <filename> - Gets the filename from the DFTP server");
	puts("PUT <user's name> - Puts you name in a file on the DFTP server");
	puts("COMMENT <user's comment> - Puts a comment in a file on the DFTP server");
}


// DFTP_request ---------------------------------------------------------------
void DFTP_request(char* request){
	// TODO - Connect to server
	puts(request);
}

File: DFTP.rb
#!/usr/bin/env ruby

require 'socket'

# Creates a connection with host:port and 
# gets the response from sending request
def DFTP_request(hostname, port, request)
  s = TCPSocket.open(hostname, port)
  s.print(request + "\r\n\r\n")
  
  response = s.read
end

# Helper to get input
def get_request
  print "command > "
  gets.chomp
end

# Load server host, port, and acquire IP
host = ARGV.shift || "unyama.cs.ucdavis.edu"
port = ARGV.shift || 10000
ip = Socket::getaddrinfo(host, port)[0][3]
STDIN.flush

# Welcome Message
puts "Welcome to the simple DFTP client."
puts "Connecting to #{host}(#{ip}) on port #{port}."

# Main loop
request = get_request()
while(true)
  case request
  when "quit"
    puts "Goodbye."
    exit
  when "help"
    puts "help - Prints this message"
    puts "quit - Exits the program"
    puts "GET <filename> - Gets the filename from the DFTP server"
    puts "PUT <user's name> - Puts you name in a file on the DFTP server"
    puts "COMMENT <user's comment> - Puts a comment in a file on the DFTP server"
  else
    puts DFTP_request(host, port, request)
  end
  request = get_request()
end